### PR TITLE
Add ray casting via Kernel12 edge-triangle intersection

### PR DESCRIPTION
**Pull request by zmerlynn**
## Summary

Implement `Manifold::RayCast(origin, endpoint)` and `Manifold::RayCastAll(origin, endpoint)` by constructing a minimal single-edge `Impl` for the ray segment and calling `Kernel12` from boolean3 directly. This reuses the full symbolic perturbation machinery (Shadow01, Kernel02, Kernel11, Kernel12) for watertight results at shared edges and vertices, with zero new intersection math.

The ray `Impl` has two vertices (origin, endpoint), two halfedges forming one edge, and zero normals — so the ray contributes nothing to perturbation tiebreakers, and consistency at shared edges/vertices depends entirely on the mesh's own normals. Candidate triangles are filtered via the BVH using the ray's AABB.

- **`RayCast`** returns the nearest hit as a `RayHit` (triangle index, parametric distance `[0, 1]`, 3D position, face normal). Returns `faceID == -1` on miss.
- **`RayCastAll`** returns all hits sorted by distance.

Both share a `RayCastImpl` helper that builds the ray Impl and runs the BVH query.

## Follow-up work

- **Bindings**: C API (`ManifoldRayHit`, `manifold_ray_cast`), Python (`.ray_cast()`), WASM (`.rayCast()`)
- **Convenience overload**: `RayCast(origin, direction, maxDist)` for the direction + distance form

Partial implementation for #1640

## Test plan

- [x] 19 new tests: RayCast (14) + RayCastAll (5) covering all three axes, diagonal rays, sphere, nearest-hit, two-object traversal, empty manifold, zero-length ray, watertight edge/vertex, normals, hit positions
- [x] All 358 tests pass (326 existing + 19 new + 13 from other merged PRs)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by zmerlynn** [comment]
Added `RayCastAll` since implementing it required refactoring `RayCast` into a shared `RayCastImpl` helper anyway — the only difference is accumulate-all vs track-best, so it made more sense to include both than to churn the same code twice.

File: include/manifold/common.h [hunk 1]
```diff
@@ -154,6 +154,22 @@ struct Smoothness {
   double smoothness;
 };
 
+/**
+ * @brief Result of a ray cast query against a Manifold.
+ */
+struct RayHit {
+  /// The triangle index that was hit, or -1 for no hit.
+  int faceID = -1;
```
Comment (elalish) [inline include/manifold/common.h]: I think this should be a `uint64_t` so as to ensure it can handle anything `MeshGL64` throws at it.

File: include/manifold/manifold.h [hunk 2]
```diff
@@ -395,6 +395,13 @@ class Manifold {
   double MinGap(const Manifold& other, double searchLength) const;
   ///@}
 
+  /** @name Spatial Queries
+   */
+  ///@{
+  RayHit RayCast(vec3 origin, vec3 endpoint) const;
```
Comment (elalish) [inline include/manifold/manifold.h]: I think I prefer your original that returned `std::optional<RayHit>` - any reason you changed it?

File: src/boolean3.cpp [hunk 3]
```diff
@@ -528,4 +528,107 @@ Boolean3::Boolean3(const Manifold::Impl& inP, const Manifold::Impl& inQ,
   }
 #endif
 }
+namespace {
```
Comment (elalish) [inline src/boolean3.cpp]: Don't we already have an anonymous namespace in this file somewhere? In any case this namespace is generally at the top of the file.

File: src/boolean3.cpp [hunk 4]
```diff
@@ -528,4 +528,107 @@ Boolean3::Boolean3(const Manifold::Impl& inP, const Manifold::Impl& inQ,
...
+template <typename RecordHit>
+void RayCastImpl(const Manifold::Impl& mesh, vec3 origin, vec3 endpoint,
+                 RecordHit& recordHit) {
+  using Impl = Manifold::Impl;
+  // Build a minimal single-edge Impl representing the ray segment.
+  // Kernel12 treats inA as an edge mesh, so we need two vertices, two
+  // halfedges (forward + backward), and one face normal. Zero vertex normals
+  // and face normal mean the ray contributes nothing to perturbation
+  // tiebreakers — consistency at shared edges/vertices depends entirely on
+  // the mesh's own normals.
```
Comment (elalish) [inline src/boolean3.cpp:542]: Sounds good, but make sure you've checked the tests with ASAN, just in case we didn't notice something being accessed out of bounds.

File: src/boolean3.cpp [hunk 5]
```diff
@@ -528,4 +528,107 @@ Boolean3::Boolean3(const Manifold::Impl& inP, const Manifold::Impl& inQ,
...
+      bestTri = tri;
+    }
+  };
+  RayCastImpl(*this, origin, endpoint, recordHit);
+
+  if (bestTri < 0) return {};
+  RayHit hit;
+  hit.faceID = bestTri;
+  hit.distance = bestT;
+  hit.position = origin + bestT * dir;
```
Comment (elalish) [inline src/boolean3.cpp]: This introduces rounding error - it would be better to use the position calculated by Kernel12.

File: test/manifold_test.cpp [hunk 6]
```diff
@@ -1274,3 +1274,188 @@ TEST(Manifold, OpenscadCrash) {
...
+  EXPECT_EQ(hit.faceID, -1);
+}
+
+TEST(Manifold, RayCastNormal) {
+  Manifold cube = Manifold::Cube(vec3(1), true);
+
+  // Ray along -Z should hit bottom face, normal should point -Z
+  RayHit hit = cube.RayCast(vec3(0, 0, -5), vec3(0, 0, 5));
+  EXPECT_GE(hit.faceID, 0);
+  EXPECT_NEAR(hit.normal.z, -1.0, 1e-10);
```
Comment (elalish) [inline test/manifold_test.cpp]: EXPECT_FLOAT_EQ?

File: test/manifold_test.cpp [hunk 7]
```diff
@@ -1274,3 +1274,188 @@ TEST(Manifold, OpenscadCrash) {
...
+}
+
+TEST(Manifold, RayCastSphere) {
+  Manifold sphere = Manifold::Sphere(1.0, 32);
+
+  // Through center
+  RayHit hit = sphere.RayCast(vec3(0, 0, -5), vec3(0, 0, 5));
+  EXPECT_GE(hit.faceID, 0);
+  EXPECT_NEAR(la::length(hit.position), 1.0, 0.05);  // near the surface
```
Comment (elalish) [inline test/manifold_test.cpp]: Seems like a pretty wide tolerance - can it get a bit tighter?

File: test/manifold_test.cpp [hunk 8]
```diff
@@ -1274,3 +1274,188 @@ TEST(Manifold, OpenscadCrash) {
...
+}
+
+TEST(Manifold, RayCastWatertightEdge) {
+  // Cast two rays that bracket a shared edge of a cube. Both should hit.
+  // If edge handling were inconsistent, one might miss.
+  Manifold cube = Manifold::Cube(vec3(1), true);
+
+  // Centered unit cube: the diagonal edge of the bottom face passes through
+  // Y=0 (approximately). Rays just above and below should both hit.
+  RayHit hitA = cube.RayCast(vec3(0.25, 1e-15, -5), vec3(0.25, 1e-15, 5));
```
Comment (elalish) [inline test/manifold_test.cpp]: Why [0.25, 0]? With a centered cube, won't the diagonal go right through [0, 0] instead?

File: test/manifold_test.cpp [hunk 9]
```diff
@@ -1274,3 +1274,188 @@ TEST(Manifold, OpenscadCrash) {
...
+  EXPECT_GE(hitB.faceID, 0);
+
+  // And exactly on the edge
+  RayHit hitEdge = cube.RayCast(vec3(0.25, 0.0, -5), vec3(0.25, 0.0, 5));
+  EXPECT_GE(hitEdge.faceID, 0);
+}
+
+TEST(Manifold, RayCastWatertightVertex) {
+  // Ray exactly through a vertex of the cube. The symbolic perturbation
+  // should assign the hit to exactly one triangle.
```
Comment (elalish) [inline test/manifold_test.cpp]: If I'm not mistaken, you actually tested these cases in the first few tests, right? Maybe just move these comments / names up there.

File: test/manifold_test.cpp [hunk 10]
```diff
@@ -1274,3 +1274,188 @@ TEST(Manifold, OpenscadCrash) {
...
+
+  auto hits = cube.RayCastAll(vec3(0, 0, -5), vec3(0, 0, 5));
+  ASSERT_EQ(hits.size(), 2);
+  // Sorted by distance
+  EXPECT_LT(hits[0].distance, hits[1].distance);
+  EXPECT_NEAR(hits[0].position.z, -0.5, 1e-10);
+  EXPECT_NEAR(hits[1].position.z, 0.5, 1e-10);
+}
+
+TEST(Manifold, RayCastAllMiss) {
```
Comment (elalish) [inline test/manifold_test.cpp]: Might be good to add a test of interesting an edge near-exactly at the silhouette, you can test getting two intersections or zero - never one.

**Review by elalish — CHANGES_REQUESTED** [review]
Thanks, looking good!

File: include/manifold/common.h (hunk 1 above)
Comment (zmerlynn) [reply include/manifold/common.h]: Done, switched to `uint64_t`.

File: include/manifold/manifold.h (hunk 2 above)
Comment (zmerlynn) [reply include/manifold/manifold.h]: Good call — switched back to `std::optional<RayHit>`.

File: src/boolean3.cpp (hunk 3 above)
Comment (zmerlynn) [reply src/boolean3.cpp]: Inlined the logic into `RayCastAll` directly, so no separate namespace needed.

File: src/boolean3.cpp (hunk 4 above)
Comment (zmerlynn) [reply src/boolean3.cpp:542]: Clean under CI-matching ASAN build (`MANIFOLD_DEBUG=ON`, `MANIFOLD_STRICT=ON`, `clang++`).

File: src/boolean3.cpp (hunk 5 above)
Comment (zmerlynn) [reply src/boolean3.cpp]: Good catch — using Kernel12's computed position `v` directly now.

File: test/manifold_test.cpp (hunk 6 above)
Comment (zmerlynn) [reply test/manifold_test.cpp]: Updated to `EXPECT_FLOAT_EQ` throughout.

File: test/manifold_test.cpp (hunk 7 above)
Comment (zmerlynn) [reply test/manifold_test.cpp]: Bumped to 128-segment sphere with `1e-4` tolerance.

File: test/manifold_test.cpp (hunk 8 above)
Comment (zmerlynn) [reply test/manifold_test.cpp]: You're right — moved into `RayCastHitCube` which casts through `(0,0)`.

File: test/manifold_test.cpp (hunk 9 above)
Comment (zmerlynn) [reply test/manifold_test.cpp]: Merged the edge case into `RayCastHitCube`. Kept the vertex test separate since it's a distinct degenerate case.

File: test/manifold_test.cpp (hunk 10 above)
Comment (zmerlynn) [reply test/manifold_test.cpp]: Added — verifies 0 or 2 hits at the silhouette edge, never 1.

**Comment by zmerlynn** [comment]
`RayCast` now delegates to `RayCastAll().front()` — `Collider::Collisions` visits all AABB-overlapping leaves regardless so there's no early termination to exploit, and this keeps the code in one place.

File: src/boolean3.cpp [hunk 11]
```diff
@@ -528,4 +528,77 @@ Boolean3::Boolean3(const Manifold::Impl& inP, const Manifold::Impl& inQ,
   }
 #endif
 }
+std::optional<RayHit> Manifold::Impl::RayCast(vec3 origin,
+                                              vec3 endpoint) const {
+  // Delegates to RayCastAll since Collider::Collisions visits all
+  // AABB-overlapping leaves regardless — there's no early termination
+  // to exploit, so the only overhead is a small vector and sort.
```
Comment (elalish) [inline src/boolean3.cpp]: I think this is a good call, and given that, let's just keep the API simple and remove this one and rename `RayCastAll` to `RayCast` (referring both to these internal methods and the public API).

**Review by elalish — CHANGES_REQUESTED** [review]
Looking great - if you can just fix the conflicts and simplify the API, I think we're about ready to merge.

File: src/boolean3.cpp (hunk 11 above)
Comment (zmerlynn) [reply src/boolean3.cpp]: Done — dropped the optional variant and renamed RayCastAll → RayCast.

**Comment by zmerlynn** [comment]
Simplified the API per your suggestion: dropped the `optional`-returning `RayCast`, renamed `RayCastAll` → `RayCast`. Single function returning `std::vector<RayHit>`, sorted by distance. Since `Collider::Collisions` visits all AABB-overlapping leaves regardless, there's no early-termination advantage to a single-hit variant.

**Review by elalish — APPROVED** [review]
Thanks!
